### ISSUE TITLE
imagenet resnet pytorch slow steps/second

### ISSUE THREAD

**Issue by priyakasimbeg** [mlcommons/algorithmic-efficiency#212]
The imagenet resent pytorch workload seems to be running at least twice as slow as the jax workload

## Description
The graph below shows the jax and pytorch run. They both ran for 31h but the pytorch run only completed about 100K steps.
<img width="820" alt="image" src="https://user-images.githubusercontent.com/12614254/197079882-3a5f3aae-d41b-45da-aa69-1f4e5a414fc8.png">

*Provide more details on the bug itself*
On kasimbeg-4 ran:
```
torchrun --redirects 1:0,2:0,3:0,4:0,5:0,6:0,7:0 --standalone --nnodes=1 --nproc_per_node=8 submission_runner.py \
    --framework=pytorch \
    --data_dir=~/local_data/imagenet/pytorch \
    --imagenet_v2_data_dir=~ \
    --experiment_dir=~ \
    --experiment_name=target_setting \
    --workload=imagenet_resnet \
    --submission_path=reference_algorithms/target_setting_algorithms/pytorch_nesterov.py \
    --tuning_search_space=reference_algorithms/target_setting_algorithms/imagenet_resnet/tuning_search_space.json > results_imagenet_resnet_pytorch.txt 2>&1
```

**Comment by priyakasimbeg** [comment]
Ah ok yes, I will rerun both w the update flag for about 12 h and add the result.

**Comment by pomonam** [comment]
Moreover, now with the checkpoints implemented, you can simply re-run the code and it will start from the final points.

**Comment by runame** [comment]
Thanks for looking into this! Let me know if I can help with anything regarding investigating the speed difference.

**Comment by pomonam** [comment]
In my initial experiments, it seems like the data loader is causing the most discrepancy. Below is the output given by Pytorch (note that I only ran the code for an 0.5h):
```
----------------------------------------------------------------------------------------------------------------------------------------------------------
|  Action                       |  Mean Duration (s)    |  Std Duration (s)     |  Num Calls            |  Total Time (s)       |  Percentage %         |
----------------------------------------------------------------------------------------------------------------------------------------------------------
|  Total                        |  -----                |  -----                |  7504                 |  1952.0               |  100 %                |
----------------------------------------------------------------------------------------------------------------------------------------------------------
|  Train                        |  1950.2               |  0.0                  |  1                    |  1950.2               |  99.907               |
|  Update parameters            |  0.36978              |  0.072261             |  3748                 |  1386.0               |  71.003               |
|  Data selection               |  0.079406             |  0.079867             |  3748                 |  297.61               |  15.247               |
|  Evaluation                   |  32.191               |  7.9381               |  4                    |  128.76               |  6.5965               |
|  Initializing dataset         |  6.2116               |  0.0                  |  1                    |  6.2116               |  0.31822              |
|  Initializing model           |  2.5319               |  0.0                  |  1                    |  2.5319               |  0.12971              |
|  Initializing optimizer       |  0.12336              |  0.0                  |  1                    |  0.12336              |  0.0063196            |
----------------------------------------------------------------------------------------------------------------------------------------------------------
```
and the below is by Jax:
```
----------------------------------------------------------------------------------------------------------------------------------------------------------
|  Action                       |  Mean Duration (s)    |  Std Duration (s)     |  Num Calls            |  Total Time (s)       |  Percentage %         |
----------------------------------------------------------------------------------------------------------------------------------------------------------
|  Total                        |  -----                |  -----                |  10652                |  1954.4               |  100 %                |
----------------------------------------------------------------------------------------------------------------------------------------------------------
|  Train                        |  1949.4               |  0.0                  |  1                    |  1949.4               |  99.742               |
|  Update parameters            |  0.3337               |  0.51689              |  5322                 |  1776.0               |  90.867               |
|  Data selection               |  0.0044586            |  0.066115             |  5322                 |  23.729               |  1.2141               |
|  Evaluation                   |  21.125               |  9.5886               |  4                    |  84.499               |  4.3234               |
|  Initializing model           |  10.519               |  0.0                  |  1                    |  10.519               |  0.53823              |
|  Initializing dataset         |  1.0959               |  0.0                  |  1                    |  1.0959               |  0.056073             |
|  Initializing optimizer       |  0.89024              |  0.0                  |  1                    |  0.89024              |  0.04555              |
----------------------------------------------------------------------------------------------------------------------------------------------------------
```
I tried to change some configurations, but feel like the hardware is causing this discrepancy. Since we are storing the ImageNet data in HDD (instead of SSD), I suspect that the PyTorch dataloader is much slower.

**Comment by znado** [comment]
I'm not sure if the original post was run with my timing fixes, because if not then the pytorch checkpointing seems particularly slow and could have contributed to further discrepancies between framework timings (previously we were counting checkpointing time in the 31hr cap). In the original post pytorch only got ~100k steps compared to jax's 250k steps, which is a 60% slowdown. But the sum of the times for update params and data selection show pytorch ~25% slower (`(0.36978+0.079406)  / (0.36978+0.079406)= 0.752825`). This is ignoring the much higher variances on the jax timings.

But for the sake of debugging, lets assume that it's the data selection that is the slowdown.

Where were you storing the data in HDD? IIUC on GCP we always set up the VMs using an SSD, which we copied the data over to before running. Were you loading the data directly from GCS? If so, we should copy it first.

Regardless, should we add [FFCV](https://github.com/libffcv/ffcv) as a setup step for the pytorch imagenet datasets, and use that to speed up data loading?

**Comment by znado** [comment]
@pomonam have you use FFCV before? Would you be able to try putting that into the imagenet pytorch data pipeline to see if it can improve our speeds?

**Comment by pomonam** [comment]
I have not used it before, but I can implement it after fixing the remaining ViT issues (which I think is almost almost done).

**Comment by priyakasimbeg** [comment]
I am not sure what did the trick but the last imagenet resnet pytorch run I did actually looked good in terms of the steps per second. It got to about 150K steps in 20h. Compared to the imagenet resnet jax (which got to 250K steps in 31h) they look comparable now (8K steps per hour vs 7.5K steps per hour).

**Comment by priyakasimbeg** [comment]
Ah I think it may have been my CUDA version. I upgraded it from 11.5 to 11.7 before starting the run.

**Comment by runame** [comment]
Juhan increased the number of workers in the PyTorch data loader. This can significantly speed up the data selection step, so this also could have been the reason (and seems more likely to me). The optimal number depends on the setup, e.g. on the number of CPUs.

**Comment by priyakasimbeg** [comment]
Ok cool, I think that will definitely help. Although I started the run 20h ago and I see that https://github.com/mlcommons/algorithmic-efficiency/pull/229 w the change to nun_workers was merged 7h ago.
Also other relevant bits of info; I got sometimes got CUBLAS_STATUS_NOT_INITIALIZED w CUDA 11.5 which was also fixed by upgrading to 11.7. Perhaps, something weird was happening under the hood when it did work with 11.5.

**Comment by runame** [comment]
Ah I missed that. Would be very interesting if changing the CUDA version caused the improved speed.

**Comment by pomonam** [comment]
Nice. That is good to know :). I believe that this difference can be further decreased if we use a local SSD to store the data (I attached the difference speed in IO with the current (balanced) vs SSD. This could potentially close all the gap issues we have.  

<img width="926" alt="Screen Shot 2022-10-26 at 2 50 49 PM" src="https://user-images.githubusercontent.com/34424774/198111462-8f6f9f84-5525-4972-94c9-f3ebad8ac55b.png">

**Comment by pomonam** [comment]
Note that I also upgraded to CUDA 11.7 when running the experiments too and seem to work fine, so 11.7 might be the good version to use.

**Comment by priyakasimbeg** [comment]
And also related the CUDNN version, which I upgraded from 8.3 to 8.5 too. @pomonam  are you also using cudnn 8.5?

**Comment by pomonam** [comment]
Yes, I also upgraded cudnn to 8.5 in my environment.

**Comment by priyakasimbeg** [comment]
Reran with latest fixes and the speed gap looks a lot better but still outside of tolerance (~3steps/second for jax vs ~2steps/second for pytorch).

Profiling results for jax resnet:
```
----------------------------------------------------------------------------------------------------------------------------------------------------------
|  Action                       |  Mean Duration (s)    |  Std Duration (s)     |  Num Calls            |  Total Time (s)       |  Percentage %         |
----------------------------------------------------------------------------------------------------------------------------------------------------------
|  Total                        |  -----                |  -----                |  257250               |  4.4899e+04           |  100 %                |
----------------------------------------------------------------------------------------------------------------------------------------------------------
|  Train                        |  4.4897e+04           |  0.0                  |  1                    |  4.4897e+04           |  99.993               |
|  Update parameters            |  0.33245              |  0.095867             |  128580               |  4.2746e+04           |  95.204               |
|  Evaluation                   |  14.914               |  2.2503               |  86                   |  1282.6               |  2.8567               |
|  Data selection               |  0.0034588            |  0.0053569            |  128580               |  444.73               |  0.99049              |
|  Initializing model           |  10.339               |  0.0                  |  1                    |  10.339               |  0.023026             |
|  Initializing dataset         |  1.177                |  0.0                  |  1                    |  1.177                |  0.0026214            |
|  Initializing optimizer       |  1.0522               |  0.0                  |  1                    |  1.0522               |  0.0023434            |
----------------------------------------------------------------------------------------------------------------------------------------------------------
```

Profiling results for pytorch resnet
```
----------------------------------------------------------------------------------------------------------------------------------------------------------
|  Action                       |  Mean Duration (s)    |  Std Duration (s)     |  Num Calls            |  Total Time (s)       |  Percentage %         |
----------------------------------------------------------------------------------------------------------------------------------------------------------
|  Total                        |  -----                |  -----                |  260265               |  6.7908e+04           |  100 %                |
----------------------------------------------------------------------------------------------------------------------------------------------------------
|  Train                        |  6.7906e+04           |  0.0                  |  1                    |  6.7906e+04           |  99.997               |
|  Update parameters            |  0.36871              |  0.030265             |  130067               |  4.7957e+04           |  70.62                |
|  Data selection               |  0.095152             |  0.098643             |  130067               |  1.2376e+04           |  18.225               |
|  Evaluation                   |  28.27                |  1.4742               |  127                  |  3590.3               |  5.2871               |
|  Initializing dataset         |  6.7378               |  0.0                  |  1                    |  6.7378               |  0.009922             |
|  Initializing model           |  2.105                |  0.0                  |  1                    |  2.105                |  0.0030997            |
|  Initializing optimizer       |  0.55151              |  0.0                  |  1                    |  0.55151              |  0.00081215           |
----------------------------------------------------------------------------------------------------------------------------------------------------------
```

**Comment by pomonam** [comment]
With the recent change @mikerabbat suggested, I was able to bridge this gap (on a 0.5h run):

PyTorch:
```
----------------------------------------------------------------------------------------------------------------------------------------------------------
|  Action                       |  Mean Duration (s)    |  Std Duration (s)     |  Num Calls            |  Total Time (s)       |  Percentage %         |
----------------------------------------------------------------------------------------------------------------------------------------------------------
|  Total                        |  -----                |  -----                |  9636                 |  1971.5               |  100 %                |
----------------------------------------------------------------------------------------------------------------------------------------------------------
|  Train                        |  1969.7               |  0.0                  |  1                    |  1969.7               |  99.905               |
|  Update parameters            |  0.33362              |  0.060603             |  4814                 |  1606.0               |  81.461               |
|  Evaluation                   |  37.14                |  7.4504               |  4                    |  148.56               |  7.5352               |
|  Data selection               |  0.0022307            |  0.04729              |  4814                 |  10.738               |  0.54468              |
|  Initializing model           |  4.3086               |  0.0                  |  1                    |  4.3086               |  0.21854              |
|  Initializing dataset         |  4.1837               |  0.0                  |  1                    |  4.1837               |  0.2122               |
|  Initializing optimizer       |  0.27054              |  0.0                  |  1                    |  0.27054              |  0.013722             |
----------------------------------------------------------------------------------------------------------------------------------------------------------
```
Jax:
```
----------------------------------------------------------------------------------------------------------------------------------------------------------
|  Action                       |  Mean Duration (s)    |  Std Duration (s)     |  Num Calls            |  Total Time (s)       |  Percentage %         |
----------------------------------------------------------------------------------------------------------------------------------------------------------
|  Total                        |  -----                |  -----                |  10652                |  1954.4               |  100 %                |
----------------------------------------------------------------------------------------------------------------------------------------------------------
|  Train                        |  1949.4               |  0.0                  |  1                    |  1949.4               |  99.742               |
|  Update parameters            |  0.3337               |  0.51689              |  5322                 |  1776.0               |  90.867               |
|  Data selection               |  0.0044586            |  0.066115             |  5322                 |  23.729               |  1.2141               |
|  Evaluation                   |  21.125               |  9.5886               |  4                    |  84.499               |  4.3234               |
|  Initializing model           |  10.519               |  0.0                  |  1                    |  10.519               |  0.53823              |
|  Initializing dataset         |  1.0959               |  0.0                  |  1                    |  1.0959               |  0.056073             |
|  Initializing optimizer       |  0.89024              |  0.0                  |  1                    |  0.89024              |  0.04555              |
----------------------------------------------------------------------------------------------------------------------------------------------------------
```

It seems like removing the sync batch norm had the largest effect. I can submit the PR or wait for the official PR.

**Comment by pomonam** [comment]
Closing with the new changes.
